src/ape_cache/query.py

```diff
@@ -0,0 +1,472 @@
+                incompatible QueryType
+
+        Returns:
+            `sqlalchemy.sql.expression.TextClause`
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @_perform_query_clause.register
+    def _perform_block_clause(self, query: BlockQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(Blocks.number >= query.start_block)
+            .where(Blocks.number <= query.stop_block)
+            .where(Blocks.number % query.step == 0)
+        )
+
+    @_perform_query_clause.register
+    def _perform_transaction_clause(self, query: BlockTransactionQuery) -> Select:
+        return select([Transactions]).where(Transactions.block_hash == query.block_id)
+
+    @_perform_query_clause.register
+    def _perform_contract_event_clause(self, query: ContractEventQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(ContractEvents.block_number >= query.start_block)
+            .where(ContractEvents.block_number <= query.stop_block)
+            .where(ContractEvents.block_number % query.step == 0)
+        )
+
+    @singledispatchmethod
+    def perform_query(self, query: QueryType) -> Optional[Iterator]:  # type: ignore
+        """
+        Performs the requested query from cache.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Raises:
+            :class:`~ape.exceptions.QueryEngineError`: When given an
+                incompatible QueryType
+
+        Returns:
+            Optional[Iterator]
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @perform_query.register
+    def _perform_block_query(self, query: BlockQuery) -> Optional[Iterator[BlockAPI]]:
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self._perform_query_clause(query))
+
+                if not result:
+                    # NOTE: Should be unreachable if estimated correctly
+                    raise QueryEngineError(f"Could not perform query:\n{query}")
+
+                for row in result:
+                    yield self.provider.network.ecosystem.decode_block(dict(row.items()))
+
+        except QueryEngineError as err:
+            logger.error(f"Database not initiated: {str(err)}")
+
+    @perform_query.register
+    def _perform_transaction_query(self, query: BlockTransactionQuery) -> Optional[Iterator[Dict]]:
+        try:
```

Comment:
Same here:
```suggestion
    def _perform_transaction_query(self, query: BlockTransactionQuery) -> Iterator[Dict]:
```

Reply:
Also unsure why `Dict` and not `TransactionAPI`?

Reply:
I cannot `decode_receipt` here because of how incompatible our table is against the ReceiptAPI. We're missing data because of TransactionAPI differences from ReceiptAPI. We just need to get the data from the database here, and there is no conflict of returning data. Now the returns won't be exactly the same, but we're returning a pandas DataFrame at the end anyway.

Reply:
> we're returning a pandas DataFrame at the end anyway.

well, that's actually not entirely true. part of the refactor to using iterators was to use the query manager in more parts of the codebase, including the non-Pandas bits

this will work for now, make an issue to refactor this table to use `ReceptAPI` instead, or if you want to give that a shot now you can

Reply:
Sounds good, will do